Unify stun and turn server urls

danxuliu opened the pull request:
Requires #5491 and #5503

~~This is a breaking change for Talk 12.~~ The settings endpoint version was bumped, so clients can still use the previous API if needed.

The format resembles the `RTCIceServer` format, so the returned values can be directly used by the WebUI like done until now. Mobile clients will need to be adjusted, though.

**Important:** The [RTCIceServer MDN page](https://developer.mozilla.org/en-US/docs/Web/API/RTCIceServer) states that `Avoid specifying an unnecessarily large number of URLs in the urls property; the startup time for your connection will go up substantially. Every server in the list will be contacted and tried out before one is selected to be used for negotiation.` It should be verified whether that applies too when using trickle ICE or not; if it does then it might be needed to rethink the format before merging this.

- **Update:** I have not been able to reproduce the described behaviour (although I have not done any fancy test like delaying packets for a specific server or something like that; I just used real servers, real servers unreachable by a specific protocol and wrong addresses). Even with several servers the connection could be established before the gathering of candidates finished. 

  Internally both [Firefox](https://hg.mozilla.org/releases/mozilla-beta/file/a9556ebc4a0470ad0b4d0d47fccb227b88089bb3/dom/media/webrtc/jsapi/MediaTransportHandler.cpp#l395) and [Chromium](https://chromium.googlesource.com/chromium/src/+/60fe159c56ce23f45fab5326719d6bb386cc434a/third_party/blink/renderer/modules/peerconnection/rtc_peer_connection.cc#356) just iterate through all the `urls` and add a separate server for each of them; in practice it seems that it does not make any difference to provide four server elements each with a single url or a single server element with four urls.

  So I guess that the statement above refers only to systems where trickle ICE is not being used :shrug: 

### Before

```
stunservers: [
  [
    url: "stunServer1"
  ],
  [
    url: "stunServer2"
  ],
  [
    url: "stunServer3"
  ]
],
turnservers: [
  [
    url: [ "turnServer1-1" ],
    urls: [ "turnServer1-1" ],
    username: "user1"
    credentials: "credentials1"
  ],
  [
    url: [ "turnServer1-2" ],
    urls: [ "turnServer1-2" ],
    username: "user1"
    credentials: "credentials1"
  ],
  [
    url: [ "turnServer2" ],
    url: [ "turnServer2" ],
    username: "user2"
    credentials: "credentials2"
  ],
]
```

### After

```
stunservers: [
  [
    urls: [ "stunServer1", "stunServer2", "stunServer3" ]
  ],
],
turnservers: [
  [
    urls: [ "turnServer1-1", "turnServer1-2" ],
    username: "user1"
    credentials: "credentials1"
  ],
  [
    url: [ "turnServer2" ],
    username: "user2"
    credentials: "credentials2"
  ],
]
```

nickvergessen commented on lib/Controller/SignalingController.php:
```diff
@@ -136,29 +136,32 @@ public function getSettings(string $apiVersion, string $token = ''): DataRespons
+			$stunUrls[] = 'stun:' . $stunServer;
 		}
+		$stun = [
+			'urls' => $stunUrls
```
Can you bump the api version for signaling and create different arrays based on the already available `$apiVersion`?

danxuliu commented:
Rebased on master and bumped API version instead of modifying the previous one.

Tested with Firefox and Chromium; it should work with Safari too... but tests are welcome ;-)

Ivansss commented:
Could we add the capability "signaling-v3" here? :)

nickvergessen commented:
Yes we should add that capability and at the same time (as said in the other PR) we can drop v1+v2 at the same time as all clients for nextcloud22 need to be updated for conversation-v4 already and there is no fallback possible.

danxuliu commented:
But if v1+v2 are dropped (I guess by [replacing `v(1|2|3)` with `v3` in _routes.php_](https://github.com/nextcloud/spreed/pull/5582/files#diff-6301937af1cc1575e9e18ecd27c4a51f0d3cf9ce12e59a14709f626dfa3ef952R65)) then there is no need to have different arrays, or there is? :thinking:

nickvergessen commented:
Exactly that was the plan? Make everything simple as you don't need to support the other things.

danxuliu commented:
But then what did you mean by [`create different arrays based on the already available $apiVersion?`](https://github.com/nextcloud/spreed/pull/5582#discussion_r627952091)?

nickvergessen commented:
Hmm maybe too many context switches 🤔

danxuliu commented:
> Hmm maybe too many context switches :thinking:

Try setting event loops :-P

Capability added and previous endpoint versions dropped.

nickvergessen reviewed (changes requested):
please update all signaling urls to v3

nickvergessen reviewed (approved):
Fixed with a fixup commit

nickvergessen commented:
Rebased after #5465  and did the fixups

danxuliu commented:
Fixed a bug with the `stunservers` array (it was an object rather than an array, I had fixed it already but it seems that I failed to push :facepalm:), moved dropping all the endpoints to the third commit rather than the second one, and removed some unneeded code now that the previous versions are not supported.
